### PR TITLE
api: add reservation API

### PR DESCRIPTION
**Pull request by saintube**
Signed-off-by: saintube <saintube@foxmail.com>

### Ⅰ. Describe what this PR does

Add  `Reservation` API

### Ⅱ. Does this pull request fix one issue?

#188 

### Ⅳ. Special notes for reviews

This API follows the design in #241 .

### CODE REVIEW

File: apis/scheduling/v1alpha1/reservation_types.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Phase",type="string",JSONPath=".status.phase",description="The phase of reservation"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:printcolumn:name="Node",type="string",JSONPath=".spec.template.spec.nodeName",priority=10
+// +kubebuilder:printcolumn:name="Ttl",type="string",JSONPath=".spec.ttl",priority=10
```
Comment (eahydra) [inline apis/scheduling/v1alpha1/reservation_types.go:110]: rename Ttl to TTL

Comment (eahydra) [reply apis/scheduling/v1alpha1/reservation_types.go:110]: how to deal with `Expires`?

File: apis/scheduling/v1alpha1/groupversion_info.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+// +groupName=scheduling.koordinator.sh
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/scheme"
+)
+
+var (
+	// GroupVersion is group version used to register these objects
```
Comment (eahydra) [inline apis/scheduling/v1alpha1/groupversion_info.go:28]: are we missing the clientset?  please try to generate the clienset

**Review by eahydra — CHANGES_REQUESTED** [review]

File: apis/scheduling/v1alpha1/reservation_types.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	// Multiple owner selectors and ANDed.
+	Owners []ReservationOwner `json:"owners,omitempty"`
+	// By default, the resources requirements of reservation (specified in `template.spec`) is filtered by whether the
+	// node has sufficient free resources (i.e. ReservationRequest <  NodeFree).
+	// When `preAllocation` is set, the scheduler will skip this validation and allow overcommitment. The scheduled
+	// reservation would be waiting to be available until free resources are sufficient.
+	PreAllocation bool `json:"preAllocation,omitempty"`
+	// Time-to-Live period for the reservation.
+	// `expires` and `ttl` are mutually exclusive. If both `ttl` and `expires` are not specified, a very
+	// long TTL will be picked as default.
```
Comment (jasonliu747) [inline apis/scheduling/v1alpha1/reservation_types.go:46]: what's the default value for TTL, 24h?

File: apis/scheduling/v1alpha1/reservation_types.go (hunk 1 above)
Comment (saintube) [reply apis/scheduling/v1alpha1/reservation_types.go]: Fixed. Tag for `Expires` added.

File: apis/scheduling/v1alpha1/groupversion_info.go (hunk 2 above)
Comment (saintube) [reply apis/scheduling/v1alpha1/groupversion_info.go:28]: Added.

File: apis/scheduling/v1alpha1/reservation_types.go (hunk 3 above)
Comment (saintube) [reply apis/scheduling/v1alpha1/reservation_types.go]: Yes. We suppose the defaults to 24h.

**Review by jasonliu747 — APPROVED** [review]
/lgtm

**Review by eahydra — APPROVED** [review]

**Pull request merged**